ADRs for post-ingestion enhancements and for preset definitions (#911)
guardian/newswires: docs/adrs/04-precomputation-of-presets.md

> # Precomputation of search presets

Reviewer: Nit, but for consistency?
```suggestion
# ADR 04: Precomputation of search presets
```

## Context

As described in the ADR on [Filter and Search](./02-filter-and-search.md), all views in Newswires will initially be implemented via a single combined search/filter mechanism. We are using Postgres for storage, which supports combining e.g. category code conditions and full text search in a single query.

In order to create a more familiar and easily adoptable experience for users, we plan to recreate many of the lists/feeds that were provided by the Fingerpost UI (e.g. 'UK News', 'Cricket scores'). Under the hood, these will be saved searches which can be combined with other user-defined filters and search terms. We've sometimes referred to these internally as 'buckets', but in the UI we've gone for 'presets'.

> The searches underlying the familiar presets are often very complex (as they were in the Fingerpost implementation) because different agencies provide different metadata and none of them map precisely onto the categories our users are interested in.As an example, we might need to add inclusions and exclusions based on category code and 'slug' text. For example, in order to differentiate Rugby League from Rugby Union, and to differentiate rugby _stories_ from rugby _results_, we need to look for `"RUGBYL" / "RUGBYU" / "RUGBY UNION"` in the slug, and exclude any of a number of category codes which are applied to results only:

Reviewer: nit (missing space):
```suggestion
The searches underlying the familiar presets are often very complex (as they were in the Fingerpost implementation) because different agencies provide different metadata and none of them map precisely onto the categories our users are interested in. As an example, we might need to add inclusions and exclusions based on category code and 'slug' text. For example, in order to differentiate Rugby League from Rugby Union, and to differentiate rugby _stories_ from rugby _results_, we need to look for `"RUGBYL" / "RUGBYU" / "RUGBY UNION"` in the slug, and exclude any of a number of category codes which are applied to results only:
```

### Precompute preset membership and persist it in the database

For each preset we have accumulated a set of conditions for membership, based on data which is available at ingestion time. Given this, we could compute which preset(s) a given story belongs to at ingestion time, and persist this in the database, e.g. via a many-to-many relation table:

| item_id (fk to item.id) | preset_id (fk to preset table.id) |
| ----------------------- | --------------------------------- |
| 1                       | 1                                 |
| 1                       | 5                                 |
| 2                       | 5                                 |

This would make the SQL query to fetch items belonging to a given preset much more straightforward, e.g.:

```sql
SELECT i.headline
FROM item i
JOIN item_presets ip ON ip.item_id = i.id
WHERE ip.preset_id = $1
ORDER BY i.first_created DESC
LIMIT 100;
```

> There are good reasons to think that this would be quicker to execute at query time than the complex queries we currently have. If it worked as expected, then it would also unlock the possibility of combining presets in a given user query (e.g. "all sports stories that aren't in the cricket presets").

Reviewer: Really minor nit:
```suggestion
There are good reasons to think that this would be quicker to execute at query time than the complex queries we currently have. If it worked as expected, then it would also unlock the possibility of combining presets in a given user query (e.g. 'all sports stories that aren't in the cricket presets').
```
